File: src/Rulesage.Composition/Extensions/DependencyInjectionExtensions.cs
using JetBrains.Annotations;
using Microsoft.Extensions.DependencyInjection;
using Rulesage.Composition.Services.Abstractions;
using Rulesage.Composition.Services.Implementations;

namespace Rulesage.Composition.Extensions;

[UsedImplicitly(ImplicitUseKindFlags.Access, ImplicitUseTargetFlags.Members)]
public static class ServiceCollectionExtensions
{
    extension(IServiceCollection collection)
    {
        public IServiceCollection AddCompositionModule()
        {
            collection.AddScoped<ICompositionContextBuilder, CompositionContextBuilder>();
            collection.AddScoped<IPlanner, Planner>();
            
            return collection;
        }
    }
}

File: src/Rulesage.Common/Extensions/DependencyInjectionExtensions.fs
namespace Rulesage.Graph.Extensions

open System.Runtime.CompilerServices
open Microsoft.Extensions.DependencyInjection
open Rulesage.Common.Repositories.Abstractions
open Rulesage.Common.Repositories.Implementations

type ServiceCollectionExtensions =
    [<Extension>]
    static member AddInfrastructure(services: IServiceCollection) =
        services
            .AddScoped<IRecordRepository, RecordRepository>()
            .AddScoped<IActionRepository, ActionRepository>()
            .AddScoped<IRuleRepository, RuleRepository>()
            .AddScoped<ICommunityRepository, CommunityRepository>()
        |> ignore

        services
